xray/usage/reportusage.go

```diff
@@ -0,0 +1,177 @@
+	minXrayVersion = "3.80.0"
+
+	xrayUsageApiPath = "api/v1/usage/events/send"
+
+	EcosystemUsageBaseUrl = "https://usage-ecosystem.jfrog.info/"
+	ecosystemUsageApiPath = "api/usage/report"
+
+	ReportUsagePrefix = "Usage Report: "
+)
+
+type ReportUsageAttribute struct {
+	AttributeName  string
+	AttributeValue string
+}
+
+func (rua *ReportUsageAttribute) isEmpty() bool {
+	return rua.AttributeName == ""
+}
+
+type ReportXrayEventData struct {
+	Attributes map[string]string `json:"data,omitempty"`
+	ProductId  string            `json:"product_name"`
+	EventId    string            `json:"event_name"`
+	Origin     string            `json:"origin,omitempty"`
+}
+
+func SendXrayUsageEvents(serviceManager xray.XrayServicesManager, events ...ReportXrayEventData) error {
+	if len(events) == 0 {
+		return errorutils.CheckErrorf(ReportUsagePrefix + "Nothing to send.")
+	}
+	config := serviceManager.Config()
+	if config == nil {
+		return errorutils.CheckErrorf(ReportUsagePrefix + "Expected full config, but no configuration exists.")
+	}
+	xrDetails := config.GetServiceDetails()
+	if xrDetails == nil {
+		return errorutils.CheckErrorf(ReportUsagePrefix + "Xray details not configured.")
+	}
+	xrayVersion, err := xrDetails.GetVersion()
+	if err != nil {
+		return errors.New(ReportUsagePrefix + "Couldn't get Xray version. Error: " + err.Error())
+	}
+	if !isVersionCompatible(xrayVersion) {
+		log.Debug(fmt.Sprintf(ReportUsagePrefix+"Expected Xray version %s or above, got %s", minXrayVersion, xrayVersion))
+		return nil
+	}
+
+	url, err := utils.BuildArtifactoryUrl(xrDetails.GetUrl(), xrayUsageApiPath, make(map[string]string))
+	if err != nil {
+		return errors.New(ReportUsagePrefix + err.Error())
+	}
+	clientDetails := xrDetails.CreateHttpClientDetails()
+
+	bodyContent, err := xrayUsageEventsToJson(events...)
+	if err != nil {
+		return errors.New(ReportUsagePrefix + err.Error())
+	}
+	utils.AddHeader("Content-Type", "application/json", &clientDetails.Headers)
+	resp, body, err := serviceManager.Client().SendPost(url, bodyContent, &clientDetails)
+	if err != nil {
+		return errors.New(ReportUsagePrefix + "Couldn't send usage info. Error: " + err.Error())
+	}
+
+	err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK, http.StatusAccepted)
+	if err != nil {
+		return err
+	}
+
+	log.Debug(ReportUsagePrefix+"Usage info sent successfully.", "Xray response:", resp.Status)
+	return nil
+}
+
+// Returns an error if the Xray version is not compatible to run usage api
+func isVersionCompatible(xrayVersion string) bool {
+	// API exists from Xray version 3.80.0 and above:
+	version := versionutil.NewVersion(xrayVersion)
+	return version.AtLeast(minXrayVersion)
+}
+
+func CreateUsageEvent(productId, featureId string, additionalAttributes ...ReportUsageAttribute) ReportXrayEventData {
+	reportInfo := ReportXrayEventData{ProductId: productId, EventId: GetExpectedXrayEventName(productId, featureId), Origin: "API"}
+
+	if len(additionalAttributes) > 0 {
```

Comment:
@yahavi taught me that it's better to use the opposite condition:

if len(additionalAttributes) = 0 {
return
}
instead of wrapping a lot of code inside an if